File: tools/sit/prepare-diffsit-inputs.js
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
import { cp, mkdir, readdir, rm, stat, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import {
  isMain,
  parseOptionArgs,
  requireOption,
  resolvePathUnder,
  writeJsonFile,
} from './common.js';

const REPORT_FILES = {
  json: 'diffsit-report.json',
  text: 'diffsit-report.txt',
  html: 'diffsit-report.html',
};

const ISSUES_FILE = 'issues.jsonl';

export function loadRules(rulesJson) {
  const rawRules = JSON.parse(rulesJson);
  if (!Array.isArray(rawRules)) {
    throw new Error('rulesJson must be a JSON array');
  }

  const fullRuleKeys = new Set();
  for (const rawRule of rawRules) {
    if (rawRule === null || typeof rawRule !== 'object' || Array.isArray(rawRule)) {
      throw new Error('each rule entry must be an object');
    }
    const rawRepository = rawRule.repository;
    const rawRuleKey = rawRule.ruleKey ?? rawRule.rule_key;
    if (typeof rawRuleKey !== 'string' || rawRuleKey.length === 0) {
      throw new Error('each rule entry must have a non-empty ruleKey');
    }
    if (rawRuleKey.includes(':')) {
      fullRuleKeys.add(normalizeFullRuleKey(rawRuleKey));
      continue;
    }
    if (typeof rawRepository !== 'string' || rawRepository.length === 0) {
      throw new Error('each rule entry must have a non-empty repository');
    }
    fullRuleKeys.add(`${rawRepository.toLowerCase()}:${rawRuleKey.toUpperCase()}`);
  }

  return [...fullRuleKeys].sort();
}

export async function prepareDiffsitInputs({
  rulesJson,
  baseDir,
  targetDir,
  reportsDir,
  manifestOutput,
}) {
  const ruleKeys = loadRules(rulesJson);
  const targetProjects = await listProjectExports(targetDir, 'target');
  const preparedBaseDir = join(dirname(manifestOutput), 'prepared-base');

  await rm(preparedBaseDir, { recursive: true, force: true });
  await mkdir(preparedBaseDir, { recursive: true });

  const baseProjects = await listBaseProjectExports(baseDir);
  for (const [projectName, projectDir] of baseProjects) {
    await copyProjectExport(projectDir, join(preparedBaseDir, projectName));
  }

  for (const [projectName, projectDir] of targetProjects) {
    if (!baseProjects.has(projectName)) {
      await synthesizeEmptyBaseline(projectDir, join(preparedBaseDir, projectName));
    }
  }

  await mkdir(reportsDir, { recursive: true });
  const manifest = {
    runs: [
      {
        name: 'sonarjs',
        rule_keys: ruleKeys,
        rule_filter: ruleKeys.join(','),
        base_dir: preparedBaseDir,
        target_dir: targetDir,
        report_dir: reportsDir,
        reports: Object.fromEntries(
          Object.entries(REPORT_FILES).map(([format, file]) => [format, join(reportsDir, file)]),
        ),
      },
    ],
  };
  await writeJsonFile(manifestOutput, manifest);
  return manifest;
}

async function listProjectExports(root, label) {
  if (!(await isDirectory(root))) {
    throw new Error(`${label} SIT export directory does not exist: ${root}`);
  }

  const entries = await readdir(root, { withFileTypes: true });
  const projectDirs = entries
    .filter(entry => entry.isDirectory() && !entry.name.startsWith('.'))
    .sort((a, b) => a.name.localeCompare(b.name));
  const missingIssueFiles = [];
  const projects = new Map();
  for (const entry of projectDirs) {
    const projectDir = join(root, entry.name);
    if (!(await isFile(join(projectDir, ISSUES_FILE)))) {
      missingIssueFiles.push(entry.name);
      continue;
    }
    projects.set(entry.name, projectDir);
  }

  if (missingIssueFiles.length > 0) {
    throw new Error(
      `${label} project export(s) missing ${ISSUES_FILE} under ${root}: ${missingIssueFiles.join(', ')}`,
    );
  }
  if (projects.size === 0) {
    throw new Error(`${label} SIT export directory contains no project exports: ${root}`);
  }
  return projects;
}

async function listBaseProjectExports(root) {
  if (!(await isDirectory(root))) {
    return new Map();
  }

  const projects = new Map();
  const entries = await readdir(root, { withFileTypes: true });
  for (const entry of entries
    .filter(entry => entry.isDirectory() && !entry.name.startsWith('.'))
    .sort((a, b) => a.name.localeCompare(b.name))) {
    const projectDir = join(root, entry.name);
    if (!(await isFile(join(projectDir, ISSUES_FILE)))) {
      console.warn(
        `Warning: base project export missing ${ISSUES_FILE}, treating as empty baseline: ${projectDir}`,
      );
      continue;
    }
    projects.set(entry.name, projectDir);
  }
  return projects;
}

async function copyProjectExport(source, destination) {
  await rm(destination, { recursive: true, force: true });
  await cp(source, destination, { recursive: true });
}

async function synthesizeEmptyBaseline(targetProjectDir, destination) {
  await rm(destination, { recursive: true, force: true });
  await mkdir(destination, { recursive: true });
  await writeFile(join(destination, ISSUES_FILE), '', 'utf8');
  const metadataPath = join(targetProjectDir, 'metadata.json');
  if (await isFile(metadataPath)) {
    await cp(metadataPath, join(destination, 'metadata.json'));
  }
}

async function isDirectory(path) {
  try {
    return (await stat(path)).isDirectory();
  } catch (error) {
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}

async function isFile(path) {
  try {
    return (await stat(path)).isFile();
  } catch (error) {
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}

function normalizeFullRuleKey(ruleKey) {
  const [repository, key] = ruleKey.split(':', 2);
  if (!repository || !key) {
    throw new Error(`invalid rule key: ${ruleKey}`);
  }
  return `${repository.toLowerCase()}:${key.toUpperCase()}`;
}

async function main() {
  const args = parseOptionArgs(process.argv);
  const cwd = process.cwd();
  const manifest = await prepareDiffsitInputs({
    rulesJson: requireOption(args, '--rules-json'),
    baseDir: resolvePathUnder(cwd, requireOption(args, '--base-dir'), '--base-dir'),
    targetDir: resolvePathUnder(cwd, requireOption(args, '--target-dir'), '--target-dir'),
    reportsDir: resolvePathUnder(cwd, requireOption(args, '--reports-dir'), '--reports-dir'),
    manifestOutput: resolvePathUnder(
      cwd,
      requireOption(args, '--manifest-output'),
      '--manifest-output',
    ),
  });
  console.log(`Wrote DiffSIT manifest (${manifest.runs.length} run(s))`);
}

if (isMain(import.meta.url)) {
  main().catch(error => {
    console.error(error instanceof Error ? error.message : error);
    process.exit(1);
  });
}

File: dev-tools/scripts/sit_pr_diffsit/prepare_diffsit_inputs.py
"""Prepare language-scoped DiffSIT inputs and a run manifest.

The workflow downloads target and baseline SIT exports from the current run.
This script aligns those directories so DiffSIT can compare one language at a
time without Python invoking DiffSIT.
"""

import argparse
import json
import shutil
import sys
from collections import defaultdict
from pathlib import Path
from typing import Any

if __package__ in {None, ""}:
    sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from sit_pr_fps.common import resolve_path_under

REPORT_FILES = {
    "json": "diffsit-report.json",
    "text": "diffsit-report.txt",
    "html": "diffsit-report.html",
}
ISSUES_FILE = "issues.jsonl"


def path_under_cwd(label: str):
    def parse_path(value: str) -> Path:
        try:
            return resolve_path_under(Path.cwd(), value, label)
        except ValueError as err:
            raise argparse.ArgumentTypeError(str(err)) from err

    return parse_path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Prepare DiffSIT inputs for the SIT PR workflow")
    parser.add_argument("--rules-json", required=True)
    parser.add_argument("--base-dir", required=True, type=path_under_cwd("--base-dir"))
    parser.add_argument("--target-dir", required=True, type=path_under_cwd("--target-dir"))
    parser.add_argument("--reports-dir", required=True, type=path_under_cwd("--reports-dir"))
    parser.add_argument("--manifest-output", required=True, type=path_under_cwd("--manifest-output"))
    parser.add_argument(
        "--rule-key-format",
        default="{language}dre:{rule_id}",
        help=(
            "Python format string for full rule keys used in the DiffSIT "
            "rule-filter (default keeps sonar-skunk's DRE convention)."
        ),
    )
    parser.add_argument(
        "--flat-bundles",
        action="store_true",
        help=(
            "If set, treat --base-dir and --target-dir as flat bundle "
            "directories (no language sub-layer)."
        ),
    )
    return parser.parse_args()


def load_rules(rules_json: str) -> dict[str, list[str]]:
    try:
        raw_rules = json.loads(rules_json)
    except json.JSONDecodeError as err:
        raise ValueError(f"rules_json must be valid JSON: {err}") from err

    if not isinstance(raw_rules, list):
        raise ValueError("rules_json must be a JSON array")

    grouped: dict[str, set[str]] = defaultdict(set)
    for raw_rule in raw_rules:
        if not isinstance(raw_rule, dict):
            raise ValueError("each rules_json entry must be an object")

        language = raw_rule.get("language")
        rule_id = raw_rule.get("rule_id")
        if not isinstance(language, str) or not language:
            raise ValueError("each rules_json entry must have a non-empty string language")
        if not isinstance(rule_id, str) or not rule_id:
            raise ValueError("each rules_json entry must have a non-empty string rule_id")

        grouped[language.lower()].add(rule_id.upper())

    return {language: sorted(rule_keys) for language, rule_keys in sorted(grouped.items())}


def exported_rule_key(language: str, rule_key: str, rule_key_format: str) -> str:
    return rule_key_format.format(language=language, rule_id=rule_key)


def visible_directories(root: Path) -> list[Path]:
    return [
        path for path in sorted(root.iterdir()) if path.is_dir() and not path.name.startswith(".")
    ]


def list_project_exports(language_dir: Path, label: str) -> dict[str, Path]:
    if not language_dir.is_dir():
        raise FileNotFoundError(f"{label} language export directory does not exist: {language_dir}")

    project_dirs = visible_directories(language_dir)
    missing_issue_files = [
        project_dir.name for project_dir in project_dirs if not (project_dir / ISSUES_FILE).is_file()
    ]
    if missing_issue_files:
        missing_projects = ", ".join(missing_issue_files)
        raise FileNotFoundError(
            f"{label} project export(s) missing {ISSUES_FILE} under {language_dir}: {missing_projects}"
        )

    projects = {project_dir.name: project_dir for project_dir in project_dirs}
    if not projects:
        raise FileNotFoundError(f"{label} language export directory contains no project exports: {language_dir}")
    return projects


def copy_project_export(source: Path, destination: Path) -> None:
    if destination.exists():
        shutil.rmtree(destination)
    shutil.copytree(source, destination)


def synthesize_empty_baseline(target_project_dir: Path, destination: Path) -> None:
    if destination.exists():
        shutil.rmtree(destination)
    destination.mkdir(parents=True, exist_ok=True)
    (destination / ISSUES_FILE).write_text("", encoding="utf-8")

    target_metadata = target_project_dir / "metadata.json"
    if target_metadata.is_file():
        shutil.copy2(target_metadata, destination / "metadata.json")


def prepare_language_base(base_dir: Path, target_dir: Path, prepared_base_dir: Path, language: str) -> None:
    target_language_dir = target_dir / language
    target_projects = list_project_exports(target_language_dir, "target")

    prepared_language_base = prepared_base_dir / language
    if prepared_language_base.exists():
        shutil.rmtree(prepared_language_base)
    prepared_language_base.mkdir(parents=True, exist_ok=True)

    base_language_dir = base_dir / language
    base_projects: dict[str, Path] = {}
    if base_language_dir.is_dir():
        for project_dir in visible_directories(base_language_dir):
            if not (project_dir / ISSUES_FILE).is_file():
                print(
                    f"Warning: base project export missing {ISSUES_FILE}, "
                    f"treating as empty baseline: {project_dir}",
                    file=sys.stderr,
                )
                continue
            base_projects[project_dir.name] = project_dir

    for project_name, base_project_dir in base_projects.items():
        copy_project_export(base_project_dir, prepared_language_base / project_name)

    for project_name, target_project_dir in target_projects.items():
        if project_name not in base_projects:
            synthesize_empty_baseline(target_project_dir, prepared_language_base / project_name)


def build_manifest(
    grouped_rules: dict[str, list[str]],
    target_dir: Path,
    reports_dir: Path,
    prepared_base_dir: Path,
    rule_key_format: str,
    flat_bundles: bool,
) -> dict[str, Any]:
    runs = []
    for language, rule_keys in grouped_rules.items():
        language_reports_dir = reports_dir / language
        base_dir_path = prepared_base_dir if flat_bundles else prepared_base_dir / language
        target_dir_path = target_dir if flat_bundles else target_dir / language
        runs.append(
            {
                "language": language,
                "rule_keys": rule_keys,
                "rule_filter": ",".join(
                    exported_rule_key(language, rule_key, rule_key_format) for rule_key in rule_keys
                ),
                "base_dir": str(base_dir_path),
                "target_dir": str(target_dir_path),
                "report_dir": str(language_reports_dir),
                "reports": {
                    report_format: str(language_reports_dir / filename)
                    for report_format, filename in REPORT_FILES.items()
                },
            }
        )
    return {"runs": runs}


def prepare_inputs(
    *,
    rules_json: str,
    base_dir: Path,
    target_dir: Path,
    reports_dir: Path,
    manifest_output: Path,
    rule_key_format: str,
    flat_bundles: bool,
) -> dict[str, Any]:
    grouped_rules = load_rules(rules_json)
    prepared_base_dir = manifest_output.parent / "prepared-base"

    if flat_bundles:
        prepare_flat_base(base_dir, target_dir, prepared_base_dir)
    else:
        for language in grouped_rules:
            prepare_language_base(base_dir, target_dir, prepared_base_dir, language)

    for language in grouped_rules:
        (reports_dir / language).mkdir(parents=True, exist_ok=True)

    manifest = build_manifest(
        grouped_rules, target_dir, reports_dir, prepared_base_dir, rule_key_format, flat_bundles
    )
    manifest_output.parent.mkdir(parents=True, exist_ok=True)
    manifest_output.write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")
    return manifest


def prepare_flat_base(base_dir: Path, target_dir: Path, prepared_base_dir: Path) -> None:
    target_projects = list_project_exports(target_dir, "target")

    if prepared_base_dir.exists():
        shutil.rmtree(prepared_base_dir)
    prepared_base_dir.mkdir(parents=True, exist_ok=True)

    base_projects: dict[str, Path] = {}
    if base_dir.is_dir():
        for project_dir in visible_directories(base_dir):
            if (project_dir / ISSUES_FILE).is_file():
                base_projects[project_dir.name] = project_dir

    for project_name, base_project_dir in base_projects.items():
        copy_project_export(base_project_dir, prepared_base_dir / project_name)

    for project_name, target_project_dir in target_projects.items():
        if project_name not in base_projects:
            synthesize_empty_baseline(target_project_dir, prepared_base_dir / project_name)


def main() -> int:
    args = parse_args()
    try:
        manifest = prepare_inputs(
            rules_json=args.rules_json,
            base_dir=args.base_dir,
            target_dir=args.target_dir,
            reports_dir=args.reports_dir,
            manifest_output=args.manifest_output,
            rule_key_format=args.rule_key_format,
            flat_bundles=args.flat_bundles,
        )
    except (FileNotFoundError, ValueError) as err:
        print(f"Error: {err}", file=sys.stderr)
        return 1

    print(f"Wrote DiffSIT manifest to {args.manifest_output} ({len(manifest['runs'])} run(s))")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
